### PR TITLE
[Exec] Adding collection result consumer type 

### PR DESCRIPTION
**Pull request by ramtinms**
This PR introduces collection execution result consumer, which would facilitate message passing between future of computation, post-processing and query engines. 
In the future PRs, the trie commit, proof generation, provider, ... logic would move covered by consumers.

### CODE REVIEW

File: engine/execution/computation/result/consumer.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	// BlockHeader returns the block header in which collection was included
+	BlockHeader() *flow.Header
+
+	// Collection returns the content of the collection
+	Collection() *flow.Collection
+
+	// RegisterUpdates returns all registers that were updated during collection execution
+	UpdatedRegisters() flow.RegisterEntries
+
+	// TouchedRegisters returns all registers that has been touched (read/written) during collection execution
```
Comment (pattyshack) [inline engine/execution/computation/result/consumer.go:18]: can we move away from tracking the touch registers, and instead move toward tracking reads from the underlying storage?  you only the read set to re-execute the transactions

File: engine/execution/messages.go [hunk 2]
```diff
@@ -52,3 +52,70 @@ func NewEmptyComputationResult(
...
+			}
+			endTxnIndex := cr.TransactionResultIndex[colIndex]
+			return cr.TransactionResults[startTxnIndex:endTxnIndex]
+		},
+	}
+
+}
+
+type ColResSnapshot struct {
+	BlockHeaderFunc        func() *flow.Header
```
Comment (pattyshack) [inline engine/execution/messages.go:91]: why not just implement them as methods?

File: engine/execution/computation/computer/result_collector.go [hunk 3]
```diff
@@ -274,6 +279,14 @@ func (collector *resultCollector) processTransactionResult(
 		return nil
 	}
 
+	// TODO(ramtin): trigger consumers concurrently
```
Comment (pattyshack) [inline engine/execution/computation/computer/result_collector.go:282]: result collector is already pipeline.  most of the time, it's stall waiting for transaction results. we probably won't benefit much from additional concurrency (except for the last collection)

File: engine/execution/computation/computer/result_collector.go [hunk 4]
```diff
@@ -274,6 +279,14 @@ func (collector *resultCollector) processTransactionResult(
...
+	for _, consumer := range collector.consumers {
```
Comment (pattyshack) [inline engine/execution/computation/computer/result_collector.go:283]: move this loop into commitCollection()

File: engine/execution/computation/result/consumer.go (hunk 1 above)
Comment (ramtinms) [reply engine/execution/computation/result/consumer.go:18]: That's not true, as of now you need the full register touches to verify execution on the verification nodes, given we have to provide proofs for the registers that are going to be written as well.

Comment (pattyshack) [reply engine/execution/computation/result/consumer.go:18]: the union of read set + updated set gives you the touch set.  there's no need to make the touch set first class

File: engine/execution/computation/computer/result_collector.go (hunk 4 above)
Comment (ramtinms) [reply engine/execution/computation/computer/result_collector.go:283]: Ah, nice, I wrote this before commitCollection was here :D, moving it to the commitCollection

File: engine/execution/computation/computer/result_collector.go (hunk 3 above)
Comment (ramtinms) [reply engine/execution/computation/computer/result_collector.go:282]: from TPS perspective yes, but might be consumer latency they could benefit from faster calls, e.g. provider or bitswap make things available to the outside world faster and not waiting for other consumers

File: engine/execution/computation/computer/result_collector.go (hunk 4 above)
Comment (ramtinms) [reply engine/execution/computation/computer/result_collector.go]: updated at [066683c](https://github.com/onflow/flow-go/pull/4049/commits/066683c63c4de690855a8e6443d9a51c769e12dc)

File: engine/execution/computation/computer/result_collector.go (hunk 3 above)
Comment (ramtinms) [reply engine/execution/computation/computer/result_collector.go]: removed the comment for now 
[066683c](https://github.com/onflow/flow-go/pull/4049/commits/066683c63c4de690855a8e6443d9a51c769e12dc)

File: engine/execution/messages.go (hunk 2 above)
Comment (ramtinms) [reply engine/execution/messages.go]: I had some thoughts initially but not valid anymore, so just changed it back
[066683c](https://github.com/onflow/flow-go/pull/4049/commits/066683c63c4de690855a8e6443d9a51c769e12dc)

File: engine/execution/computation/result/consumer.go (hunk 1 above)
Comment (ramtinms) [reply engine/execution/computation/result/consumer.go]: yeah we could keep the API on primitives, changed it here [066683c](https://github.com/onflow/flow-go/pull/4049/commits/066683c63c4de690855a8e6443d9a51c769e12dc)

**Review by pattyshack — APPROVED** [review]

File: engine/execution/messages.go [hunk 5]
```diff
@@ -52,3 +52,61 @@ func NewEmptyComputationResult(
 		},
 	}
 }
+
+func (cr ComputationResult) TxResByColIndex(colIndex int) []flow.TransactionResult {
```
Comment (janezpodhostnik) [inline engine/execution/messages.go:56]: This would be more readable with a full name, or at least a comment.

**Review by janezpodhostnik — APPROVED** [review]

File: engine/execution/messages.go (hunk 5 above)
Comment (ramtinms) [reply engine/execution/messages.go]: updated the name, also made it private

File: engine/execution/computation/computer/result_collector.go [hunk 6]
```diff
@@ -226,6 +231,13 @@ func (collector *resultCollector) commitCollection(
 		NumberOfCollections: 1,
 	}
 
+	for _, consumer := range collector.consumers {
+		err = consumer.OnExecutedCollection(collector.result.CollectionResult(collection.collectionIndex))
```
Comment (zhangchiqing) [inline engine/execution/computation/computer/result_collector.go:235]: ⚠️  `CollectionResult` might return `nil` here, I'm not sure if `OnExecutedCollection` is able to handle `nil` value. 

I would suggest to let `CollectionResult` return `*ColResSnapshot, bool` instead.

File: module/mempool/entity/executableblock.go [hunk 7]
```diff
@@ -86,6 +86,15 @@ func (b *ExecutableBlock) Collections() []*CompleteCollection {
 	return collections
 }
 
+// CollectionAt returns an address to a collection at the given index,
+// if index out of range, nil will be returned
+func (b *ExecutableBlock) CollectionAt(index int) *CompleteCollection {
+	if index < 0 && index > len(b.Block.Payload.Guarantees) {
+		return nil
```
Comment (zhangchiqing) [inline module/mempool/entity/executableblock.go:93]: Same here, I would suggest to return 
```
return nil, fmt.Errorf("index %v out of range [0, %v]", index, len(b.Block.Payload.Guarantees)
```

**Pull request merged**